### PR TITLE
generation of multiple blueprint specifications

### PR DESCRIPTION
**Pull request by shlomitk1**
Signed-off-by: SHLOMITK@il.ibm.com <shlomitk@il.ibm.com>

In addition to the module selection, decision is made in what geography the module will run based on user requirements and the existing set of M4DModules. For each geography, an appropriate blueprint specification will be generated.
This is an initial version that does not work with cluster metadata information. Thus, the blueprints/modules are generated per a geography and not a specific cluster.

### CODE REVIEW

**Comment by simanadler** [comment]
@shlomitk1 I understand this is a first step, and thus you simplified the problem to distributing per geography.  However, is there an underlying assumption that there will be only one cluster per geography? If so, I don't think this is something we should assume.  Perhaps starting with the assumption that the clusterID=geography and coding using clusterID would result in less re-write in the next PR?

File: manager/controllers/app/modules/moduleinterface.go [hunk 1]
```diff
@@ -70,90 +69,116 @@ func (m *Selector) GetError() string {
...
+	geo := m.selectGeography(item)
 	// append moduleinstances to the list
 	instances = append(instances, ModuleInstanceSpec{
-		AssetID: item.AssetID,
-		Module:  m.GetModule(),
-		Args:    args,
+		AssetID:   item.AssetID,
+		Module:    m.GetModule(),
+		Args:      args,
+		Geography: geo,
```
Comment (simanadler) [inline manager/controllers/app/modules/moduleinterface.go:78]: Is this meant to remain geo in the future or will it change to cluster ID?

File: manager/controllers/app/modules/moduleinterface.go [hunk 2]
```diff
@@ -70,90 +69,116 @@ func (m *Selector) GetError() string {
...
 	})
 	for _, dep := range m.GetDependencies() {
 		instances = append(instances, ModuleInstanceSpec{
-			AssetID: item.AssetID,
-			Module:  dep,
-			Args:    args,
+			AssetID:   item.AssetID,
+			Module:    dep,
+			Args:      args,
+			Geography: geo,
```
Comment (simanadler) [inline manager/controllers/app/modules/moduleinterface.go:85]: Same as previous comment

**Review by simanadler — COMMENTED** [review]
See comment I put in PR discussion

**Comment by shlomitk1** [comment]
> 
> 
> @shlomitk1 I understand this is a first step, and thus you simplified the problem to distributing per geography. However, is there an underlying assumption that there will be only one cluster per geography? If so, I don't think this is something we should assume. Perhaps starting with the assumption that the clusterID=geography and coding using clusterID would result in less re-write in the next PR?

This is the first step. There will be a functionality of determining clusterID based on selected geography. Also, processingGeography in m4dapplication will be replaced by workloadClusterID. Assuming N >= 1 clusters per geography.

File: manager/controllers/app/modules/moduleinterface.go (hunk 1 above)
Comment (shlomitk1) [reply manager/controllers/app/modules/moduleinterface.go:78]: Not sure. A clusterID as the first step. Perhaps, to extend to a list of possible clusterIDs to make it possible to decide later on clusterID that is common to many modules. I would start with a single clusterID.

File: manager/controllers/app/modules/moduleinterface.go (hunk 2 above)
Comment (shlomitk1) [reply manager/controllers/app/modules/moduleinterface.go:85]: CluserID as well.

**Review by simanadler — APPROVED** [review]
Approving as a first step

**Pull request merged**